### PR TITLE
Remove implicit downcasts from flutter_tools

### PR DESCRIPTION
**Pull request by jonahwilliams**
Remove 750 implicit downcasts from flutter_tools

## Description

An implicit downcast is currently allowed in the Dart language, it looks something like this:

```
dynamic x;
String y = x;
```

This is equivalent to an `as` cast - the only difference is the syntax (and at one point the generated code for AOT which the tool does not use).

```
dynamic x;
var y = x as String;
```

The problem with implicit casts arises when they are used accidentally. The most common example in forgetting to `toList` an `Iterable`. This is allowed because we perform an implicit downcast from `Iterable<T>` to `List<T>`.

```
final List<String> ys = xs.map((x) => x.toString())); // Oops no .toList()
requiresListString(ys); // Fails at runtime.
```

Most recently when pushing updates to the devicelab agent, I had to repeat the process 3 times due to implicit cast errors that I had accidentally introduced!

Since the analyzer has introduced the capability to disable implicit casts, I believe we should take advantage of this for the flutter_tool. correctness here is paramount, and while we're working to measure and improve test coverage, I don't believe it is sufficient to catch all errors.

Introducing this change exposed a handful of bugs, my personal favorite:

```
int.parse(args['some_arg'] ?? 22);
```

Furthermore, it identifies several areas of the tool that rely heavily on runtime typing being correct, namely the Context/injection system, Arguments, and the ApplicationPackage class. These should all receive follow up refactors to be made more type safe, so that the explicit casting can be removed.

## Related Issues

https://github.com/dart-lang/sdk/issues/31410

## Tests

- This is an analyzer/syntax change.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter_tools/lib/src/android/android_studio.dart [hunk 1]
```diff
@@ -297,7 +297,7 @@ class AndroidStudio implements Comparable<AndroidStudio> {
     } else {
       final ProcessResult result = processManager.runSync(<String>[javaExecutable, '-version']);
       if (result.exitCode == 0) {
-        final List<String> versionLines = result.stderr.split('\n');
+        final List<String> versionLines = result.stderr.split('\n') as List<String>;
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/android/android_studio.dart]: i'm surprised by this one

File: packages/flutter_tools/lib/src/android/gradle.dart [hunk 2]
```diff
@@ -539,12 +539,12 @@ Future<void> _buildGradleProjectV2(
 
         final String name = newFile.name;
         if (name.contains('_snapshot_') || name.endsWith('.so')) {
-          final List<int> diff = bsdiff(oldFile.content, newFile.content);
-          final int ratio = 100 * diff.length ~/ newFile.content.length;
+          final List<int> diff = bsdiff(oldFile.content as List<int>, newFile.content as List<int>);
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/android/gradle.dart]: what is oldFile.content statically?

File: packages/flutter_tools/lib/src/android/gradle.dart [hunk 3]
```diff
@@ -539,12 +539,12 @@ Future<void> _buildGradleProjectV2(
...
+          final int ratio = 100 * diff.length ~/ (newFile.content.length as int);
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/android/gradle.dart]: is length not statically an int?

File: packages/flutter_tools/lib/src/application_package.dart [hunk 4]
```diff
@@ -196,7 +196,7 @@ class AndroidApk extends ApplicationPackage {
           if (!(node is xml.XmlElement)) {
             continue;
           }
-          final xml.XmlElement xmlElement = node;
+          final xml.XmlElement xmlElement = node as xml.XmlElement;
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/application_package.dart]: replace the '!(...is...)` with `is!` and this should just work. Or wrap this in the `is` instead of bailing early. We shouldn't need both, either way.

File: packages/flutter_tools/lib/src/application_package.dart [hunk 5]
```diff
@@ -407,14 +407,14 @@ class _Element extends _Entry {
     return children.firstWhere(
         (_Entry e) => e is _Attribute && e.key.startsWith(name),
         orElse: () => null,
-    );
+    ) as _Attribute;
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/application_package.dart]: you can factor out the `e is _Attribute` using `whereType` and that'll get rid of the `as`

File: packages/flutter_tools/lib/src/base/build.dart [hunk 6]
```diff
@@ -459,8 +458,9 @@ class JITSnapshotter {
           // But if engine snapshot exists, its content must match the engine snapshot
           // in baseline APK. Otherwise, we're trying to build an update at an engine
           // version that might be binary incompatible with baseline APK.
-          final Function contentEquals = const ListEquality<int>().equals;
-          if (!contentEquals(f.readAsBytesSync(), af.content)) {
+          print(af.content.runtimeType);
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/base/build.dart]: stray prints

File: packages/flutter_tools/lib/src/base/fingerprint.dart [hunk 7]
```diff
@@ -134,14 +134,15 @@ class Fingerprint {
...
-  bool operator==(dynamic other) {
+  bool operator==(Object other) {
     if (identical(other, this))
       return true;
-    if (other.runtimeType != runtimeType)
-      return false;
-    final Fingerprint typedOther = other;
-    return _equalMaps(typedOther._checksums, _checksums)
-        && _equalMaps(typedOther._properties, _properties);
+    if (other is Fingerprint) {
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/base/fingerprint.dart]: we should compare runtimeType not use `is`, in an `==` operator; see style guide

File: packages/flutter_tools/lib/src/base/os.dart [hunk 8]
```diff
@@ -177,7 +177,7 @@ class _WindowsUtils extends OperatingSystemUtils {
       if (entity is! File) {
         continue;
       }
-      final File file = entity;
+      final File file = entity as File;
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/base/os.dart]: hm, this is similar to the case above. i guess i was wrong about `is!`. does it let you avoid the `as` if you wrap instead of bailing? maybe file a bug on the sdk for this, the compiler should be able to determine that `entity` is a `File` here.

File: packages/flutter_tools/lib/src/base/os.dart [hunk 9]
```diff
@@ -249,7 +249,7 @@ class _WindowsUtils extends OperatingSystemUtils {
       final ProcessResult result = processManager.runSync(
           <String>['ver'], runInShell: true);
       if (result.exitCode == 0)
-        _name = result.stdout.trim();
+        _name = result.stdout.trim() as String;
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/base/os.dart]: i'd rather put the `as` on the `stdout` here.

File: packages/flutter_tools/lib/src/commands/analyze_base.dart [hunk 10]
```diff
@@ -167,8 +167,8 @@ class PackageDependencyTracker {
         // we are analyzing the actual canonical source for this package;
         // make sure we remember that, in case all the packages are actually
         // pointing elsewhere somehow.
-        final yaml.YamlMap pubSpecYaml = yaml.loadYaml(fs.file(pubSpecYamlPath).readAsStringSync());
-        final String packageName = pubSpecYaml['name'];
+        final yaml.YamlMap pubSpecYaml = yaml.loadYaml(fs.file(pubSpecYamlPath).readAsStringSync()) as yaml.YamlMap;
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/analyze_base.dart]: that's sad

File: packages/flutter_tools/lib/src/commands/emulators.dart [hunk 11]
```diff
@@ -44,9 +44,9 @@ class EmulatorsCommand extends FlutterCommand {
     }
 
     if (argResults.wasParsed('launch')) {
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/emulators.dart]: i reviewed up to here

**Comment by Hixie** [comment]
To make this easier to review it might make sense to do it in bits, first the ones that involve `Context`, then the ones that involve Json, etc.

File: packages/flutter_tools/lib/src/android/android_workflow.dart [hunk 12]
```diff
@@ -220,7 +220,7 @@ class AndroidLicenseValidator extends DoctorValidator {
     try {
       final ProcessResult result = await processManager.run(<String>[javaBinary, '-version']);
       if (result.exitCode == 0) {
-        final List<String> versionLines = result.stderr.split('\n');
+        final List<String> versionLines = result.stderr.split('\n') as List<String>;
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/android/android_workflow.dart]: dang, what was this without the cast?!

File: packages/flutter_tools/test/fuchsia/fuchsa_device_test.dart [hunk 13]
```diff
@@ -209,7 +209,7 @@ void main() {
   });
 
   group(FuchsiaIsolateDiscoveryProtocol, () {
-    Future<Uri> findUri(List<MockFlutterView> views, String expectedIsolateName) {
+    Future<Uri> findUri(List<MockFlutterView> views, String expectedIsolateName) async {
```
Comment (christopherfujino) [inline packages/flutter_tools/test/fuchsia/fuchsa_device_test.dart]: Why is this marked `async`? I thought you only needed `async` if you used `await`. Is it because of `.thenAnswer`?

**Review by christopherfujino — CHANGES_REQUESTED** [review]
TL;DR if the print statements were intentional, LGTM.

**Comment by christopherfujino** [comment]
add2app macos strange failure:

```
	45  XCTest                              0x0000000124d23ccb __27-[XCTestSuite performTest:]_block_invoke + 365
	46  XCTest                              0x0000000124d234a3 -[XCTestSuite _performProtectedSectionForTest:testSection:] + 55
	47  XCTest                              0x0000000124d23766 -[XCTestSuite performTest:] + 296
	48  XCTest                              0x0000000124d6c1a2 -[XCTest runTest] + 57
	49  XCTest                              0x0000000124d9be86 __44-[XCTTestRunSession runTestsAndReturnError:]_block_invoke + 171
	50  XCTest                              0x0000000124d9bfa7 __44-[XCTTestRunSession runTestsAndReturnError:]_block_invoke.80 + 68
	51  XCTest                              0x0000000124d3bbc1 -[XCTestObservationCenter _observeTestExecutionForBlock:] + 585
	52  XCTest                              0x0000000124d9bbfa -[XCTTestRunSession runTestsAndReturnError:] + 623
	53  XCTest                              0x0000000124d086b6 -[XCTestDriver runTestsAndReturnError:] + 422
	54  XCTest                              0x0000000124d8c9cd _XCTestMain + 1478
	55  libXCTestBundleInject.dylib         0x000000010f1c1bb8 __copy_helper_block_ + 0
	56  CoreFoundation                      0x00000001121be62c __CFRUNLOOP_IS_CALLING_OUT_TO_A_BLOCK__ + 12
	57  CoreFoundation                      0x00000001121bdde0 __CFRunLoopDoBlocks + 336
	58  CoreFoundation                      0x00000001121b8654 __CFRunLoopRun + 1284
	59  CoreFoundation                      0x00000001121b7e11 CFRunLoopRunSpecific + 625
	60  GraphicsServices                    0x0000000115fab1dd GSEventRunModal + 62
	61  UIKitCore                           0x000000011b1b581d UIApplicationMain + 140
	62  ios_add2app                         0x000000010ef91450 main + 112
	63  libdyld.dylib                       0x0000000115218575 start + 1
)
Test Case '-[FlutterTests testHybridView]' failed (7.755 seconds).
Test Suite 'FlutterTests' failed at 2019-04-15 11:18:48.429.
	 Executed 3 tests, with 6 failures (6 unexpected) in 27.094 (27.127) seconds
Test Suite 'ViewControllerRelease' started at 2019-04-15 11:18:48.431
Test Case '-[ViewControllerRelease testReleaseFlutterViewController]' started.
Test Case '-[ViewControllerRelease testReleaseFlutterViewController]' passed (3.616 seconds).
Test Suite 'ViewControllerRelease' passed at 2019-04-15 11:18:52.050.
	 Executed 1 test, with 0 failures (0 unexpected) in 3.616 (3.619) seconds
Test Suite 'ios_add2appTests.xctest' failed at 2019-04-15 11:18:52.052.
	 Executed 4 tests, with 6 failures (6 unexpected) in 30.710 (30.752) seconds
Test Suite 'All tests' failed at 2019-04-15 11:18:52.054.
	 Executed 4 tests, with 6 failures (6 unexpected) in 30.710 (30.757) seconds


Test session results and logs:
	/Users/anka/Library/Developer/Xcode/DerivedData/ios_add2app-csiwhdwugybmwjeggtidemdjpolj/Logs/Test/Test-ios_add2appTests-2019.04.15_11-17-03--0700.xcresult

2019-04-15 11:19:16.724 xcodebuild[14376:33144] [MT] IDETestOperationsObserverDebug: 110.034 elapsed -- Testing started completed.
2019-04-15 11:19:16.724 xcodebuild[14376:33144] [MT] IDETestOperationsObserverDebug: 0.000 sec, +0.000 sec -- start
2019-04-15 11:19:16.724 xcodebuild[14376:33144] [MT] IDETestOperationsObserverDebug: 110.034 sec, +110.034 sec -- end
Failing tests:
	-[FlutterTests testDualFlutterView]
	-[FlutterTests testDualFlutterView]
	-[FlutterTests testFullScreenCanPop]
	-[FlutterTests testFullScreenCanPop]
	-[FlutterTests testHybridView]
	-[FlutterTests testHybridView]
** TEST FAILED **

Testing started on 'iPhone X'
🕐 ELAPSED TIME: 0:03:44.631328 for ../../../build_and_test.sh  in dev/integration_tests/ios_add2app: 
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
ERROR: Last command exited with 65 (expected: zero).
Command: ../../../build_and_test.sh 
Relative working directory: dev/integration_tests/ios_add2app
```

**Comment by jonahwilliams** [comment]
Unfortunately add2app-macos is super flaky

File: packages/flutter_tools/lib/src/android/android_studio.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/android/android_studio.dart]: All of the File/Process APIs are typed as dynamic :(

File: packages/flutter_tools/lib/src/android/gradle.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/android/gradle.dart]: dynamic

File: packages/flutter_tools/lib/src/android/gradle.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/android/gradle.dart]: Like above, all of the file/devfs/process APIs use dynamic

File: packages/flutter_tools/lib/src/base/fingerprint.dart (hunk 7 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/base/fingerprint.dart]: comparing runtimeType doesn't do a type promotion, so we need to cast regardless. I'd rather have the type safety of the `is` check, and I think we should allow the tool to diverge from the framework in this regard - especially since there are only a handful of types that implement `operator.==` here.

File: packages/flutter_tools/lib/src/base/os.dart (hunk 8 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/base/os.dart]: The canonical issue is https://github.com/dart-lang/sdk/issues/25565. better type promotion is a requirement of NNBD, and last I heard was a work in progress

File: packages/flutter_tools/lib/src/android/android_workflow.dart (hunk 12 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/android/android_workflow.dart]: stderr is dynamic unfortunately.

File: packages/flutter_tools/lib/src/application_package.dart (hunk 4 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/application_package.dart]: Mentioned below: `is!` does not promote

File: packages/flutter_tools/lib/src/base/os.dart (hunk 9 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/base/os.dart]: Done

File: packages/flutter_tools/lib/src/base/build.dart (hunk 6 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/base/build.dart]: Fixed

[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/base/build.dart]: Fixed

File: packages/flutter_tools/lib/src/application_package.dart (hunk 5 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/application_package.dart]: Done, here and below

File: packages/flutter_tools/lib/src/commands/drive.dart [hunk 14]
```diff
@@ -82,7 +82,7 @@ class DriveCommand extends RunCommandBase {
 
   Device _device;
   Device get device => _device;
-  bool get shouldBuild => argResults['build'];
+  bool get shouldBuild => argResults['build'] as bool;
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/drive.dart]: cool, was just checking for this :+1:

**Comment by jonahwilliams** [comment]
>To make this easier to review it might make sense to do it in bits, first the ones that involve Context, then the ones that involve Json, etc.

I'm happy to split this up in a way that can be aided by tools, such as a directory based approach?

splitting it up based on subject matter is going to require me verifying every cast multiple times

File: packages/flutter_tools/test/fuchsia/fuchsa_device_test.dart (hunk 13 above)
Comment (jonahwilliams) [reply packages/flutter_tools/test/fuchsia/fuchsa_device_test.dart]: This needs to return a `Future<Uri>` but was currently returning only a `Uri`. The most straightforward way to fix that is to use an `async method.

Comment (zanderso) [reply packages/flutter_tools/test/fuchsia/fuchsa_device_test.dart]: You can also return `Future.value(discovertyProtocol.uri)`

**Comment by jonahwilliams** [comment]
Closing in favor of smaller refactors

File: packages/flutter_tools/lib/src/base/fingerprint.dart (hunk 7 above)
Comment (Hixie) [reply packages/flutter_tools/lib/src/base/fingerprint.dart]: the problem is it'll break when you subclass Fingerprint, and it'll break when someone copies and pastes this code into other code where it isn't as safe.

We're going to have to work out how to make this type safe in any case.

File: packages/flutter_tools/lib/src/application_package.dart (hunk 4 above)
Comment (Hixie) [reply packages/flutter_tools/lib/src/application_package.dart]: well we should still use `is!` instead of `!(...is...)` in any case. :-)

File: packages/flutter_tools/lib/src/android/gradle.dart (hunk 3 above)
Comment (Hixie) [reply packages/flutter_tools/lib/src/android/gradle.dart]: sure but the most type-safe solution here would be to cast the `content`, not the `length`.
